### ISSUE TITLE
Escoria 3 is not compatible with Godot's current master

### ISSUE THREAD

**Issue by StraToN** [godotengine/escoria#43]
My recents tries to fix #41 and https://github.com/flossmanualsfr/escoria/issues/14 have been stopped by errors during execution.

Many old functions are still called from some scripts:
- `AudioStreamPlayer.set_volume()` instead of `AudioStreamPlayer.set_volume_db()` : https://github.com/godotengine/escoria/blob/master/device/globals/bg_music.gd#L30
This one is very strange, since `set_volume` actually never existed, but no crash happened until now.

- `InputEvent` : if event.type == InputEvent.MOUSE_BUTTON && event.pressed: : https://github.com/godotengine/escoria/blob/master/device/globals/background_area.gd#L8 fails with error "type attribute does not exist"
- And probably many more.

I still use flossmanual's Escoria in Daiza for testing this.

**Comment by fleskesvor** [comment]
> AudioStreamPlayer.set_volume() instead of AudioStreamPlayer.set_volume_db() : https://github.com/godotengine/escoria/blob/master/device/globals/bg_music.gd#L30
This one is very strange, since set_volume actually never existed, but no crash happened until now.

This was discussed briefly in https://github.com/godotengine/escoria/issues/34#issuecomment-344438520, where it was also mentioned that would fail at line 28 too, where `set_loop()` is called, but that never happened, since the "stream" node was never found. Did that behaviour change in a recent build of Godot or did you implement a fix in Escoria?

> InputEvent : if event.type == InputEvent.MOUSE_BUTTON && event.pressed: : https://github.com/godotengine/escoria/blob/master/device/globals/background_area.gd#L8 fails with error "type attribute does not exist"

I also changed https://github.com/godotengine/escoria/blob/master/device/globals/background_area.gd#L34 from `connect("gui_input", self, "input")` to `connect("input_event", self, "input")` to make background areas work with the Daïza fork.

**Comment by StraToN** [comment]
>>AudioStreamPlayer.set_volume() instead of AudioStreamPlayer.set_volume_db() : https://github.com/godotengine/escoria/blob/master/device/globals/bg_music.gd#L30
>>This one is very strange, since set_volume actually never existed, but no crash happened until now.
>
>This was discussed briefly in #34 (comment), where it was also mentioned that would fail at line 28 too, where set_loop() is called, but that never happened, since the "stream" node was never found. Did that behaviour change in a recent build of Godot or did you implement a fix in Escoria?

I suppose the debugger has been improved then, because the current master of Godot complains about this inexisting function even before getting to this point at execution.
I did nothing with this script. However, I cleared the music file set inf the AudioStreamPlayer node, which may be the cause.



>>InputEvent : if event.type == InputEvent.MOUSE_BUTTON && event.pressed: : https://github.com/godotengine/escoria/blob/master/device/globals/background_area.gd#L8 fails with error "type attribute does not exist"
>
>I also changed https://github.com/godotengine/escoria/blob/master/device/globals/background_area.gd#L34 from connect("gui_input", self, "input") to connect("input_event", self, "input") to make background areas work with the Daïza fork.

Yes, I was using your branch so I could observe the same thing as you. I pushed my fixes on my own branch : https://github.com/StraToN/escoria-in-daiza/tree/remove-pnm-files.
It is odd, it seems I didn't rebase my branch on yours.

**Comment by fleskesvor** [comment]
Then you might also have seen the workaround for instanced scenes not having a filename in this commit:

https://github.com/StraToN/escoria-in-daiza/commit/998827bf9b663611a72da5b86ac7b8937a41fcb9#diff-7ba6a4f70a1502b64a4c21c0091edb60R448

The problem the lack of a filename is causing is that autosaves don't work, since the filename can't be written to the save file, which in turn makes it impossible to load the right scene. The workaround mitigates the issue, but it's no good having to resort to workarounds if it can be fixed properly.

I'm not sure if it's related to incompatibilities with the current master though, since I'm not clear on how the resource loading works/is supposed to work, so it might not be related to this issue. If not, this comment might be copied into a new issue.

**Comment by StraToN** [comment]
I see. I didn't reproduce the autosave problem yet since I am having other errors.
I am having a problem with the dialog with chameleon. It seems dd_player.gd is unable to retrieve the resource "default". Although it is correctly set, `get_resource_list()` returns `[]`.

All these problems make me want to remake Escoria from (almost) scratch. This would probably fix many of our problems, and the code would probably be cleaner, which would fix #10 . :/

**Comment by fleskesvor** [comment]
Yeah, I have to admit that I find that tempting as well. Some of the code which is very loosely coupled to the engine, like the .esc script parsing, looks solid to me, but I think there's a good chance that things like input handling and resource loading have changed so much since Escoria was conceived that at this point, maintaining it as it is means having to deal with a lot of cruft which might not even serve a purpose anymore.

**Comment by StraToN** [comment]
Indeed, that was the point behind my "almost" word choice :)
The .esc parser/compiler, as well as the global_vm principle must be kept, even though a huge rewriting is definitely necessary. I'll create an issue.

**Comment by fleskesvor** [comment]
I forgot to answer this question:

> Also, I was wondering why this code was removed? fleskesvor/escoria@fbe49d3#diff-66535240158c865227d5be39cb5bfa43L20

It's because the automatic shape logic isn't working with Godot 3, and I'm not sure how to fix it (yet). That said, I also think it needs some improvements, if it's to be kept, as I mentioned in https://github.com/flossmanualsfr/escoria/issues/12#issuecomment-353035958: 

> I wrote that logic, but in retrospect, I'm not very happy with how it works, since it makes assumptions about how your scene is structured, while you might possibly want to eg. place the Area2D with the click shape as a child of the background, so that you can save the whole background scene separately and re-use it in multiple rooms in the game.

The reason why I think it might be desireable to re-use background scenes is that it makes it easier to use the same location again but with different "props" (NPCs, items, etc.), without having to use a bunch of flags to determine which "props" are displayed etc., but instead create a different scene in the editor. Plus, the scene would also be a different scene in movie terminology, so it might be easier conceptually to a game designer.

**Comment by StraToN** [comment]
Ok, I agree on your point. I have 2 examples of this in The Secret of Monkey Island, Melee Island/.
- SCUMM bar scene. When the game begins it is full of actors ; later in the game the exact same scene is empty and decor animations are different.
![image](https://user-images.githubusercontent.com/1289207/34780216-ccbb787a-f622-11e7-8f4a-e8315b104bb1.png)
![image](https://user-images.githubusercontent.com/1289207/34780182-b58aa176-f622-11e7-928a-11a20a8dbe5c.png)


- Streets scene. At the beginning streets are populated by the group of 3 pirates and the strange guy selling maps. Also random people walk through the street from door to door. At the end of the game, the exact same scene contains pirates ghosts to confront.
![image](https://user-images.githubusercontent.com/1289207/34779724-581b48fc-f621-11e7-960d-92e59adcd1a5.png)
![image](https://user-images.githubusercontent.com/1289207/34779872-bb5d42ee-f621-11e7-9f07-f4b1e30e2840.png)


That goes in favor of a rewrite again. We should indeed make this kind of things easy to do.

**Comment by fleskesvor** [comment]
Yes, those are very good examples of what I was thinking of. Another one that comes to mind is the different Snuckey's locations (both inside and outside) in Sam & Max Hit the Road, which are mostly the same, except for some minor changes.

**Comment by mjtorn** [comment]
Sorry if this warrants a new issue, but is the work in progress the reason for getting errors?

```
error is Errors in file res://game/fallbacks.esc

Invalid action talk in fallbacks.
```

**Comment by fleskesvor** [comment]
@mjtorn That is an expected error if neither the .esc file for the item you're interacting with, nor the fallbacks.esc file has a "talk" action defined. By default, the fallbacks file only contains fallbacks for "pick_up" and "use":

    :use
    debug "use fallback!"
    say player "I can't use that."

    :pick_up
    debug "pick_up fallback!"
    say player "I can't take this up."

**Comment by mjtorn** [comment]
I'll get back to this asap, hopefully like tomorrow, but I was
testing the example tutorial and was surprised to see errors.

I spent quite a bit of time on Unity and its Adventure Creator
asset (and some others) so I shouldn't expect to have to have
everything figured out on the first day I see Godot and Escoria ;)

There are some commits in `master` to improve compatibility
with Godot 3. How stable should it be or am I better off with
the `2.1` branch for now? What's the relationship between
that and the rewrite branch?

Asking because Unity on Linux, and some assets that are not
Adventure Creator, have been causing me headache and I'd really
like a more open and text-based approach :)

Thanks!

**Comment by fleskesvor** [comment]
There was a big effort put into the framework to make it compatible with Godot 3.0 during the beta, but there have been some breaking API changes since, so I'm not sure how compatible it is at the moment. @StraToN would know the status, but I believe he's busy with GodotCon at the moment, so might take some time to get back to you.

It would be great to be able to offer something on par with AGS, Adventure Creator and Visionaire eventually, but the framework isn't quite there yet. I think we should be able to implement several new features and a better demo project in the coming months though. I agree that it's odd that by default, there are only fallback dialogue for two verbs, while the demo has six verbs (I think). Ideally, all should be covered, so as to not cause unnecessary friction for developers exploring the project. The main reason it is so is probably that the framework hasn't been used much by people other than its contributors, so feedback "from the outside" is certainly welcome.

For this error specifically, you can fix it by adding the lines

    talk:
    say player "I have nothing to say to that"

or something similar to the fallbacks.esc file, and similarly with the other verbs.

**Comment by mjtorn** [comment]
These are not the best of news, that Escoria is not on par. However, one thing
that I found annoying with Adventure Creator is the dialogue system, and to
replace it there's Dialogue System, which isn't great either. Escoria's system
for dialogue looks like what I want, just focus on writing and express the
dialogue tree that way.

I'm trying to figure out what a killer feature in Adventure Creator has been
for me but all I can think of is how annoying it sometimes is to integrate it.

Do you have anything in mind that would be big in AC but on the Escoria
roadmap?

One thing I'd want is a Full Throttle style verb coin, is that doable?
(I don't think AC does this out of the box in any way but I'm sure I saw an
example on the forums)

I'll also figure out GUIs and menus ASAP. They're not super complex by nature
but a good looking menu is sometimes hard to do in AC.

But yeah, as for the specific problem, I'll look into it ASAP when I'm not
at home, thanks! :)

**Comment by fleskesvor** [comment]
Yeah, you're right, and I think we can get that sorted out soon, so that it's at least possible to run games from the editor again. Then at least it won't be a blocker for game development.

Full Throttle style verb coins are already possible. I think it's documented in the Daïza tutorial how you can do that, but there's no example implementation there. [Amnesia Adventure](https://github.com/Double-Fine-Game-Club/amnesia-adventure) has a couple of example implementations, but the project itself is very rough; especially when it comes to project structure (and I'm to blame for much of that).

As for future features, a couple of things that come to mind, since I've made proof-of-concept implementations in Godot 2.1:

- costumes: multiple AnimationPlayers on a player on an item to be able to switch between them in esc scripts. Eg. Guybrush Threepwood's dress in LeChuck's revenge, but there might be use cases for items as well.
- multiple playable characters: this will be easier to implement in a good way after the suggested rewrite of Escoria (#44).

Some other things that have been discussed:

- gamepad support
- "exits": special kinds of items that trigger a customisable action when a player or an item enters/exits its collision shape. Can be used to trigger a scene switch, but also play an animation or trigger other game logic.

If there are other features you'd like to see added, you're very welcome to open an issue.

**Comment by mjtorn** [comment]
I actually didn't get any code investigation done today, but I did read up
more on the documentation and found some Youtube videos. Also purchased
Dog Mendonca, it looks really good and probably showcases
a bunch of features. Now I need time to play ;)

The verb coin stuff actually seems documented, I just need to get my focus
together on this.

Multiple characters is trendy for sure, but not relevant to me now. Changing
costumes on player(s) isn't that important either.

Gamepads are for sure important. I, or the missus and I, have a Steam Machine
hooked to the TV and a pad is the most convenient way. Godot also allegedly
supports PS4 builds, with additional licensing hassle, so maybe someone
may want to use Escoria to get a game out for the PS4 as well.

On the subject of exits, maybe it's not trivial, but wouldn't it be possible
to do that already, not just that easily?

Thanks for your answers and I could open issues for the ... well ... issues.

**Comment by fleskesvor** [comment]
> Yeah, you're right, and I think we can get that sorted out soon, so that it's at least possible to run games from the editor again.

After checking, this is already possible. I just pulled ab253d994b904f09916cfeaf4f0b355be30d3d78 from upstream, and it's working with the official Godot 3.0 downloads. I tested by running the demo scene with some interactions.

Gamepad support is something that should be significantly easier to support once the proposed architectural changes have been implemented. The way I see that working, is that you'd need a gamepad controllable cursor (that should also be customizable, like the action menu), that should behave pretty much like a mouse cursor, with mouse-hints, know that the current item under the cursor is the one to be interacted with, trigger walk on button press, etc. It should also be possible to use this cursor in place of the desktop mouse cursor for mouse controls if a developer wants to enable that.

Be sure to open issues for the features that are most important to you, as that will increase the chance of them becoming implemented soon.

**Comment by StraToN** [comment]
Thanks a lot @fleskesvor for answering while I was away: you are right I was indeed very busy during the FOSDEM and GodotCon so I didn't even have the time to read my mails.
You almost said everything actually. The current state of Escoria is that current master branch is mostly compatible with Godot 3, but there are probably some problems and bug that we didn't see and/or others that we just cannot fix easily yet.

This is why, we are currently in reflection about trying to remake Escoria from scratch (#44) so that we can provide more than a collection of template scene and scripts, but also specific tools for Godot to ease the work for the specific adventure/point'n'click genre (eventually including some specific ideas for RPG, visual novels and such). However this will be a huge work and we don't expect it to be ready soon.

One of our objectives is to ditch the included demo scene and replace it by a very small game introduction. I am in contact with some artists who showed interest in such an idea, so this should happen in some time - either before Escoria's remake, or most probably during/after.

**Comment by fleskesvor** [comment]
Speaking of RPGs, and I'll be brief, so as to not derail the issue too much: One thing I think might be useful for creating simple adventure RPGs, à la RPG Maker games, is to be able to use a tilemap for the terrain. Design wise, you'd probably have to use one tilemap for the terrain, which is the node you'd put the terrain script on, and one or more other tilemaps for non-walkable areas. Though, maybe one would be enough for simple locations, since you could just put collision on tiles you don't want the player to walk on. It's not something I've put a lot of thought into yet though.

**Comment by StraToN** [comment]
@fleskesvor Added to #44

**Comment by fleskesvor** [comment]
Closed by #42. Further discussion about general improvements should take place in #44 or as concrete suggestions in dedicated issues.
